keras_cv/layers/vit_layers.py

```diff
@@ -0,0 +1,204 @@
+    for Vision Transformers from:
+        - An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale
+        by Alexey Dosovitskiy et al. (https://arxiv.org/abs/2010.11929)
+
+    Based on Khalid Salama's implementation for:
+        - https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    args:
+        - project_dim: the dimensionality of the project_dim
+
+    Basic usage:
+
+    ```
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+
+    project_dim = 1024
+    num_patches = patches.shape[1] # 196
+
+    encoded_patches = keras_cv.layers.PatchEmbedding(project_dim=project_dim)(patches)
+    print(encoded_patches.shape) # (1, 197, 1024)
+    ```
+    """
+
+    def __init__(self, project_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.project_dim = project_dim
+        self.linear_projection = layers.Dense(self.project_dim)
+
+    def build(self, input_shape):
+        self.class_token = tf.random.normal([1, 1, input_shape[-1]], name="class_token")
+        self.num_patches = input_shape[1]
+        self.position_embedding = layers.Embedding(
+            input_dim=self.num_patches + 1, output_dim=self.project_dim
+        )
+
+    def call(
+        self,
+        patch,
+        interpolate=False,
+        interpolate_width=None,
+        interpolate_height=None,
+        patch_size=None,
+    ):
+        # Add learnable class token before linear projection and positional embedding
+        patch_shape = tf.shape(patch)
+        class_token_broadcast = tf.cast(
+            tf.broadcast_to(
+                self.class_token,
+                [patch_shape[0], 1, patch_shape[-1]],
+            ),
+            dtype=patch.dtype,
+        )
+        patch = tf.concat([class_token_broadcast, patch], 1)
+        # num_patches + class token
+        positions = tf.range(start=0, limit=self.num_patches + 1, delta=1)
+
+        if interpolate and None not in (
+            interpolate_width,
+            interpolate_height,
+            patch_size,
+        ):
+            encoded = self.linear_projection(
+                patch
+            ) + self.interpolate_positional_embeddings(
+                self.position_embedding(positions),
+                interpolate_width,
+                interpolate_height,
+                patch_size,
+            )
+        elif interpolate and None in (
+            interpolate_width,
+            interpolate_height,
+            patch_size,
+        ):
+            raise ValueError(
+                "`None of `interpolate_width`, `interpolate_height` and `patch_size` cannot be None if `interpolate` is True"
+            )
+        else:
+            encoded = self.linear_projection(patch) + self.position_embedding(positions)
+        return encoded
+
+    def interpolate_positional_embeddings(self, embedding, height, width, patch_size):
```

Comment:
do we expect users to call this? If not then make it private

Reply:
Nope. This is called internally when the flag is set to True. 
Prepending `__` is enough, or?

Reply:
yeah `_` is good enough

Reply:
Done!